docs/spec.md

```diff
@@ -6269,6 +6271,24 @@ continue execution without raising the corresponding status flag; similar to
 operations (e.g. complex arithmetic and certain transcendental functions) are
 implementation-defined.
 
+#### Shape mismatches
+
+StableHLO supports dynamically-shaped tensors. However, shapes have to agree at
+runtime, otherwise the behavior is undefined. StableHLO does not explicitly
+provide an op that can assert that a tensor has a given shape at runtime.
+Generating correct code is the responsibility of the producer.
+
+As a specific example, the below program is valid. However, at runtime, the
+exact shapes of `%arg0` and `%arg1` will have to be the same, otherwise the
+behavior of the program is undefined:
+
+```mlir
```

Comment:
We have the infrastructure to make a `verifyValidShapePolymorphism` pass:

```
RefineArguments
RefineShapes
CanonicalizeDynamism
VerifyNoDynamicShapes (not a pass but can walk)
```

Not sure how to best capture that here if at all